src/internal/pfsdb/branches.go

```diff
@@ -68,5 +105,173 @@ func UpsertBranch(ctx context.Context, tx *pachsql.Tx, branchInfo *pfs.BranchInf
 	).Scan(&branchID); err != nil {
 		return 0, errors.Wrap(err, "could not create branch")
 	}
+	// Compare current direct provenance to new direct provenance.
+	oldDirectProv, err := GetDirectBranchProvenance(ctx, tx, branchID)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not get direct branch provenance")
+	}
+	// Add net new direct provenance relationships.
+	toAdd := SliceDiff[string, *pfs.Branch](branchInfo.DirectProvenance, oldDirectProv, func(branch *pfs.Branch) string { return branch.Key() })
+	for _, branch := range toAdd {
+		toID, err := GetBranchID(ctx, tx, branch)
+		if err != nil {
+			return 0, errors.Wrap(err, "could not get to_id")
+		}
+		if err := CreateDirectBranchProvenance(ctx, tx, branchID, toID); err != nil {
+			return 0, errors.Wrap(err, "could not add direct branch provenance")
+		}
+	}
+	// Remove old direct provenance relationships that are no longer needed.
+	toRemove := SliceDiff[string, *pfs.Branch](oldDirectProv, branchInfo.DirectProvenance, func(branch *pfs.Branch) string { return branch.Key() })
+	for _, branch := range toRemove {
+		toID, err := GetBranchID(ctx, tx, branch)
+		if err != nil {
+			return 0, errors.Wrap(err, "could not get to_id")
+		}
+		if err := DeleteDirectBranchProvenance(ctx, tx, branchID, toID); err != nil {
+			return 0, errors.Wrap(err, "could not delete direct branch provenance")
+		}
+	}
 	return branchID, nil
 }
+
+// GetBranchID returns the unique id of a branch.
+func GetBranchID(ctx context.Context, tx *pachsql.Tx, branch *pfs.Branch) (BranchID, error) {
```

Comment:
Any thoughts about reusing the query for GetBranch() and adding the ID? It doesn't really matter, but I figure reusing queries helps down the road when we add or remove columns, its fewer things to update.

Reply:
Good idea!